keanu-project/src/main/java/io/improbable/keanu/tensor/bool/SimpleBooleanTensor.java

```diff
@@ -0,0 +1,266 @@
+    }
+
+    static BooleanTensor scalar(boolean value) {
+        return new SimpleBooleanTensor(value);
+    }
+
+    private final boolean[] data;
+    private final int[] shape;
+    private final int[] stride;
+
+    public SimpleBooleanTensor(boolean[] data, int[] shape) {
+        this.data = data;
+        this.shape = shape;
+        this.stride = TensorShape.getRowFirstStride(shape);
+    }
+
+    public SimpleBooleanTensor(boolean constant) {
+        this.data = new boolean[]{constant};
+        this.shape = Tensor.SCALAR_SHAPE;
+        this.stride = Tensor.SCALAR_STRIDE;
+    }
+
+    public SimpleBooleanTensor(int[] shape) {
+        this.data = null;
+        this.shape = shape;
+        this.stride = TensorShape.getRowFirstStride(shape);
+    }
+
+    @Override
+    public BooleanTensor and(BooleanTensor that) {
+        return duplicate().andInPlace(that);
+    }
+
+    @Override
+    public BooleanTensor or(BooleanTensor that) {
+        return duplicate().orInPlace(that);
+    }
+
+    @Override
+    public BooleanTensor not() {
+        return duplicate().notInPlace();
+    }
+
+    @Override
+    public DoubleTensor setDoubleIf(DoubleTensor trueValue, DoubleTensor falseValue) {
+        FlattenedView<Double> trueValuesFlattened = trueValue.getFlattenedView();
+        FlattenedView<Double> falseValuesFlattened = falseValue.getFlattenedView();
+
+        double[] result = new double[data.length];
+        for (int i = 0; i < result.length; i++) {
+            result[i] = data[i] ? trueValuesFlattened.getOrScalar(i) : falseValuesFlattened.getOrScalar(i);
+        }
+
+        return DoubleTensor.create(result, copyOf(shape, shape.length));
+    }
+
+    @Override
+    public IntegerTensor setIntegerIf(IntegerTensor trueValue, IntegerTensor falseValue) {
+        FlattenedView<Integer> trueValuesFlattened = trueValue.getFlattenedView();
+        FlattenedView<Integer> falseValuesFlattened = falseValue.getFlattenedView();
+
+        int[] result = new int[data.length];
+        for (int i = 0; i < result.length; i++) {
+            result[i] = data[i] ? trueValuesFlattened.getOrScalar(i) : falseValuesFlattened.getOrScalar(i);
+        }
+
+        return IntegerTensor.create(result, copyOf(shape, shape.length));
+    }
+
+    @Override
+    public BooleanTensor setBooleanIf(BooleanTensor trueValue, BooleanTensor falseValue) {
+        FlattenedView<Boolean> trueValuesFlattened = trueValue.getFlattenedView();
+        FlattenedView<Boolean> falseValuesFlattened = falseValue.getFlattenedView();
+
+        boolean[] result = new boolean[data.length];
+        for (int i = 0; i < result.length; i++) {
+            result[i] = data[i] ? trueValuesFlattened.getOrScalar(i) : falseValuesFlattened.getOrScalar(i);
+        }
+
+        return BooleanTensor.create(result, copyOf(shape, shape.length));
+    }
+
+    @Override
+    public <T> Tensor<T> setIf(Tensor<T> trueValue, Tensor<T> falseValue) {
```

Comment:
The problem comes from the primitive array type. We could make it generic but that would require an extra copy out of the boxed type to the primitive typed array. It's a lot of duplicated code but that code isn't very complex so I'm tempted to keep it. If these methods grow then it might be worth going with the generic option.

Reply:
Sweet, let's keep it